src/functions_math.jl

```diff
@@ -78,7 +78,20 @@ macro declare_matmul(MatrixT, VectorT=nothing)
     return esc(Expr(:block, codes...))
 end
 
+# The following two methods can be defined by using
+# @declare_matmul(Diagonal, AbstractVector)
+# but that overwrites existing *(1D NDA, Vector) methods
+# should improve the macro above to deal with this case
+function Base.:*(a::Diagonal, b::NamedDimsArray{<:Any, <:Any, 1})
+    return *(NamedDims.NamedDimsArray{dimnames(a)}(a), b)
+end
+
+function Base.:*(a::NamedDimsArray{<:Any, <:Any, 1}, b::Diagonal)
+    return *(a, NamedDims.NamedDimsArray{dimnames(b)}(b))
+end
+
 @declare_matmul(AbstractMatrix, AbstractVector)
+@declare_matmul(Adjoint{<:Any, <:AbstractMatrix{T1}} where T1, Adjoint{<:Any, <:AbstractVector})
```

Comment:
is the `where T1` needed to break the ambiguity?
```suggestion
@declare_matmul(Adjoint{<:Any, <:AbstractMatrix}, Adjoint{<:Any, <:AbstractVector})
```

Reply:
Yes, I believe it is needed.

`AbstractMatrix` is an alias for 'AbstractArray{T1, 2}, where T1', which puts the where T1 inside the outer curly brace.

I tried really understanding why this is needed (https://stackoverflow.com/questions/62995469/difference-between-where-keyword-inside-or-outside-the-curly-braces), but to be honest I still don't quite get why the particular form is needed here..

Reply:
That was the we have to be so specific is such a pain (again, makes me worry we'll be playing method-ambiguity-whack-a-mole forever)

Anyway, thanks for solving this particularly issue - a good one to solve